omniscidb/QueryEngine/Execute.cpp

```diff
@@ -2614,224 +2600,120 @@ std::vector<std::unique_ptr<ExecutionKernel>> Executor::createKernels(
-    // high-granularity, fragment by fragment execution instead. For scan only queries on
-    // GPU, we want the multifrag kernel path to save the overhead of allocating an output
-    // buffer per fragment.
-    auto multifrag_kernel_dispatch = [device_type,
-                                      &ra_exe_unit,
-                                      &execution_kernels,
-                                      &column_fetcher,
-                                      &co,
-                                      &eo,
-                                      &query_comp_desc,
-                                      &query_mem_desc](const int device_id,
-                                                       const FragmentsList& frag_list,
-                                                       const int64_t rowid_lookup_key) {
-      execution_kernels.emplace_back(
-          std::make_unique<ExecutionKernel>(ra_exe_unit,
-                                            device_type,
-                                            device_id,
-                                            co,
-                                            eo,
-                                            column_fetcher,
-                                            query_comp_desc,
-                                            query_mem_desc,
-                                            frag_list,
-                                            ExecutorDispatchMode::MultifragmentKernel,
-                                            rowid_lookup_key));
-    };
-    fragment_descriptor.assignFragsToMultiDispatch(multifrag_kernel_dispatch);
-  } else {
-    VLOG(1) << "Creating one execution kernel per fragment";
-    VLOG(1) << query_mem_desc.toString();
-
-    if (allow_single_frag_table_opt &&
-        (query_mem_desc.getQueryDescriptionType() == QueryDescriptionType::Projection) &&
-        table_infos.size() == 1) {
-      const auto max_frag_size =
-          table_infos.front().info.getFragmentNumTuplesUpperBound();
-      if (max_frag_size < query_mem_desc.getEntryCount()) {
-        LOG(INFO) << "Lowering scan limit from " << query_mem_desc.getEntryCount()
-                  << " to match max fragment size " << max_frag_size
-                  << " for kernel per fragment execution path.";
-        throw CompilationRetryNewScanLimit(max_frag_size);
+  for (const auto& intended_dt_itr :
+       query_comp_descs) {  // Decide on optimal dispatch mode per device type
+    // Keep in mind that the query intended for GPU may fallback to CPU!
+    // We will select dispatch modes based on conditions of actual devices, but we will
+    // modify the mode of the intended device. See usage of scheduleSingleFragment() and
+    // execution_kernels_per_device_.
+    const ExecutorDeviceType intended_dt = intended_dt_itr.first;
+    const ExecutorDeviceType actual_dt = intended_dt_itr.second->getDeviceType();
+    LOG(INFO) << "Query was inteded for " << intended_dt << ", will actually run on "
```

Comment:
What are the situations in which we would want to reschedule _some_ (not all) of the kernels to the CPU? In other words, do we need such logic at all? Maybe have a fallback policy instead?